### ISSUE TITLE
Emscripten build?

### ISSUE THREAD

**Issue by chafey** [team-charls/charls#13]
I would like to find a javascript version of a JPEG-LS decoder.  Yesterday I used emscripten to compile CharLS into JS and it works great for small images but runs out of memory on larger images.  Assuming I can fix the memory issue, this should meet my needs.  I wanted to see if the CharLS project would be interested in a PR that adds a Emscripten JS build target.  If not, I can fork and add my build step to that and keep it in sync.  Thanks

PS - The codec would be used by the cornerstone open source medical imaging sdk here:
https://github.com/chafey/cornerstone

**Comment by vbaderks** [comment]
Hi Chris,

A PR that would add a emscripten build target would be great. It would very interesting to see the asm.js performance compared to the C++ performance.

I have read about emscripten and wondered myself if an asm.js port would be faster then a native js implementation, like for example [https://github.com/rii-mango/JPEGLosslessDecoderJS]

The important thing would be to extend the CI-server scripts to ensure that they also build the  emscripten target and to have some basic tests around it.

**Comment by chafey** [comment]
Asm.js is only supported by FireFox.  Web Assembly is the new standard way for achieving native performance in the browser and will therefore be the best way to do things.  Emscripten generates fast code but brings a lot of overhead with it in terms of code size and data marshaling.  

I am going to keep working on this and will report back when I have something for you to look at.

**Comment by chafey** [comment]
Just pushed my first version here: https://github.com/chafey/charls

I included a live example that lets you load most (but not all) of the images included in the repo here:

https://rawgit.com/chafey/charls/master/emscripten/test/index.html

I am not sure how to integrate this into your main repo.  It seems like most EMSCRIPTEN projects fork from the main repo, add their EMSCRIPTEN specific stuff and keep it in sync so we might just want to follow that pattern.

**Comment by chafey** [comment]
We are looking to update the EMSCRIPTEN build of CharLS being used in CornerstoneJS - is there still interest in adding EMSCRIPTEN support to this project?  If so, I will do the work so we can stay in sync with the latest and greatest version of CharLS (ours has not been maintained and is now over 3 years old)

**Comment by vbaderks** [comment]
Yes, adding a WebAssembly build configuration to CharLS would be great.

**Comment by chafey** [comment]
OK Great, I will start working on a PR for this.  Basically it will consist of a few things:
1) Updates to the building charls wiki page
2) Additional scripts to support building the EMSCRIPTEN build - I'll create a new folder at root for this
3) Some bridge code that exposes methods to javascript for encoding and decoding JPEG-LS

You can see how I am doing this for an older version of CharLS here:
https://github.com/cornerstonejs/charls/tree/dockerbuild

**Comment by vbaderks** [comment]
Sounds good to me.

Some notes:
- I have recently made some updates to the CMakeLists.txt files (applied modern CMake rules).
- The old API JlsParameters + JpegLsDecode has been replaced with a new set of C functions charls_decoder_xxx to make it easier to extend the functionality without breaking the binary ABI every time. If you need some fine-tuning for WebAssembly please let met know. The goal is remove the old API in the next major release (scheduled for 2023)

**Comment by vbaderks** [comment]
Hi Chris, 

Do you have a timeline for your work?

I am currently thinking about creating a 2.1.0-rc1 release + a no change period of a couple weeks before releasing 2.1.0.

**Comment by chafey** [comment]
I will submit a PR by EOD Sunday Nov 17

**Comment by chafey** [comment]
FYI - just started working on this again, should have a PR ready next week

**Comment by chafey** [comment]
So after quite a bit of research and experimenting, I am now of the opinion that the JS/WASM build should be handled in a separate repository than the main C++ library.  The reason for this is that the main way to ship JS/WASM code is via NPM modules and it those have specific files/directory structures that don't mesh well with C/C++ projects.  I think it would be better/more pure to keep these things separate.  I have an example of how this would work here: 

https://github.com/chafey/modern-cpp-lib-js

The above repo is the JS/WASM build around this simple C++ library here:

https://github.com/chafey/modern-cpp-lib

The modern-cpp-lib implements a simple DICOM RLE decoder with an api similar to the CharLS decode API. 

I propose that we move forward with this approach.  Thoughts?

**Comment by chafey** [comment]
I have this working now and am seeing much faster decode times with the WASM build - 7ms to decode a 512x512x16 bit CT image on google chrome on Ubuntu 19.10 with a 6 year old iMac (Intel i7-4790k processor)

**Comment by chafey** [comment]
https://github.com/chafey/charls-js

**Comment by vbaderks** [comment]
I think your plan to move the JS/WASM to a separate repository (with CharLS as a git sub-module) is a sound plan. It makes it possible to have separate release cycles. 
I have done the same for a Windows only WIC wrapper around CharLS) and will probably also move the .NET wrapper to its own repository in the future.

I will add a link to your repository from the CharLS README.md file.

The statement "seeing much faster decode times" sounds good, but compared to what? (Plain C++ or the previous WASM build?)

**Comment by chafey** [comment]
OK - should we move the repo into team-charls organization?  I am fine either way.  In terms of speed, I was comparing it to the original EMSCRIPTEN CharLS build I did 4 years ago.  Here are some current numbers on my i7-4790k Linux machine for the same 512x512x16bit CT Image:
Native C++: 4.359 ms
WASM (Node JS 13): 7.819 ms
WASM (Google Chrome 81.0): 9.464 ms
WASM (Firefox 75.0): 8.020 ms

Browsers typically perform a bit better under Windows and Mac OS X (closer to nodejs numbers).  WASM is designed to hit 90% of native performance so we should see these numbers decrease as the WASM runtimes are optimized.

**Comment by vbaderks** [comment]
Adding your repo into the tema-charls is a good plan. Will make your member.

**Comment by chafey** [comment]
OK got that - can you give me permission to create repositories so I can transfer my charls-js repository to team-charls?

**Comment by vbaderks** [comment]
Created the empty charls-js repro and made you and admin of it.

**Comment by chafey** [comment]
Things are looking pretty good, I am hoping to get some feedback before I move it over.  
#1 Please try out the web based demo and let me know if the images look right:: https://chafey.github.io/charls-js/test/browser/index.html
#2 Any feedback on the code, tests, documentation, etc would be helpful
#3 The 16 bit grayscale images seem to come back inverted but I am not sure if this is working properly (because of the source data) or a bug somewhere

**Comment by vbaderks** [comment]
I like the index.html page, very interactive. 

It seems that 8 bit color, interleave_mode = 0 (none) gives 16 small images (T8C0E0) instead of 4.
need some time for 2 and 3.

Some initial on the website:
-WASM: is it not clearer to use the term WebAssembly?
- I would round the time values to 1 or 2 digits.
- A remark that if files are dropped on the web page, the image remains local (not uploaded to a backend server)
- There is no feedback if a bad image or a non jpeg-ls file is dropped on the web page. I would assume that the C++ exception is catched and converted to a JS exception (not sure if this is possible with the current WebAssembly architecture).

**Comment by chafey** [comment]
Thanks for the feedback, I just pushed a new version based on it:
* Changed WASM to WebAssembly
* Timing rounded to 2 digits
* Added remark about not uploading files
* Added error handling for decoding failures.  Detailed exception information is not provided - not sure it can be done easily or without slowing down decoding speed
I'll look into the interleave_mode 0 issue tomorrow. Can you clarify what you mean by "need some time for 2 and 3"?

**Comment by chafey** [comment]
@vbaderks one other thing - T8SSE0 and T8SSE3 are throwing exceptions for me "The JPEG-LS stream is encoded with a parameter value that is not supported by the CharLS decoder".  Do you know what this is about?

**Comment by vbaderks** [comment]
T8SSE0.JLS
	T8SSE3.JLS
           Image TEST8 compressed in line-interleaved mode, with near-lossless
	   error 0 and 3 (resp.), and with color components subsampled 
	   as follows:
	   R: not subsampled (256 cols x 256 rows)
	   G: subsampled 4x in the vertical direction (256 cols x 64 rows)
	   B: subsampled 2x in both directions (128 cols x 128 rows)

Support for sub-sampled pixel data has never been implemented. The JPEG-LS standard supports it and the conformance test set contains such images. 
Sub sampling is a kind of lossy compression and there has never been a use case or request for it to implement it. Except for these 2 sample images, I have never seen other implementation use it.
In general all valid JPEG-LS options that are not support by CharLS will throw this kind of exception.

**Comment by vbaderks** [comment]
@chafey,

Some more feedback on 1 and 2 and 3:

- 16 bit RGB images are not correctly displayed (I have some samples if you need them).
- Your license model is MIT. Some source files have as the SPDX-License-Identifier "BSD-3-Clause" instead of "MIT" and some source code files are missing the identifier. Recommendation from spdx.org is to apply identifier to all source code files that support comments,
- BUILDING.md states minimum is CMake 3.17, CMakeLists.txt states however: cmake_minimum_required(VERSION 3.15). CMake supports the option cmake_minimum_required(VERSION 3.15...3.17) to indicate a range of CMake options.
- DESIGN.md: useful clear text :-)
- .gitmodule: the url to charls repository is a relative path, not absolute. 
- Types.hpp contains only 1 type: FrameInfo. Is filename FrameInfo.hpp not a better filename?
- class JpegLSDecode uses int32_t. FrameInfo uses short and char for its members. My recommendation would be to use intXX_t for member variables that are intended for numbers.
-  My preference would be name the classes with noun(s): JpegLSDecode into JpegLSDecode**r** and JpegLSEncode into JpegLSEncode**r**. 
- You could also target C++17 for your jslib. CharLS is only using C++14 as upgrading is a "breaking" change. C++17 would allow to update `std::vector<unsigned char> decoded_` to `std::vector<std::byte> decoded_`
- I would copy the conformance test files into your own repository. This reduces the dependency on the location of these files in the CharLS repository. As this are static files, a copy is not a problem.
- I understand the signed/unsigned 16 bit problem for monochrome images. It is a pity that correct 16 bit images (JPEG-LS is only defined for unsigned) are now not correct displayed. Best option would be a way to indicate in the JPEG-LS bitstream that the pixels are actually signed. When the JPEG-LS bit stream is contained in a DICOM file that info comes from the DICOM metadata.

**Comment by chafey** [comment]
* 16 bit RGB - yes please share some images and I will add support 
* license - fixed, set to MIT and added to source
* CMake version set to 3.17
* DESIGN - thanks :)
* .gitmodule - relative paths were specifically recommended by the modern cmake document: https://cliutils.gitlab.io/modern-cmake/chapters/projects/submodule.html
* Types.hpp -> FrameInfo.hpp
* use of intxx_t - Fixed
* rename classes to nouns - fixed
* C++17 - fixed, but left types as unsigned char because of issue with embind
* I'll look into this soon as I want to add more DICOM images anyway
* Yes, I thought about adding code to the browset test to detect dicom and parse them but decided to not do that to keep it more "pure"

**Comment by vbaderks** [comment]
16 bit RGB samples (encoded in the 3 interleave options), actually large images.
[artificial-sample 16bit rgb.zip](https://github.com/team-charls/charls/files/4560957/artificial-sample.16bit.rgb.zip)

.gitmodule \relative path: Your right, when both projects are located on the same service (github) then relative is better. 

You could add an UI checkbox (default true) to index.html to let the user decide how to decode 16 bit monochrome images (signed or unsigned). This would mimic the way a DICOM viewer would do it (it gets that info from the Pixel Representation tag).

**Comment by chafey** [comment]
Added a checkbox for signed data.  Also added support for 16 bit RGB which works for interleaveMode 1 and 2, but not 0.  I think my logic for interleave mode 0 is right, but I am not sure - do you mind checking?  I decided to rename the repo charlsjs (removed the dash) and will be delivering it soon.  Thanks!

**Comment by chafey** [comment]
@vbaderks can you look at my logic for interleaveMode 0 here: 

https://github.com/chafey/charls-js/blob/master/test/browser/index.html#L117

I think my logic is right but its not display right

**Comment by vbaderks** [comment]
I will have a look at it. Will also double check that the image I send is correctly encoded.

**Comment by chafey** [comment]
OK great!  FYI - I added encoding and downloading to the web page, check it out:

https://chafey.github.io/charls-js/test/browser/index.html

**Comment by vbaderks** [comment]
I analyzed the 16-RGB problems. There were 3 problems:
1) The image artificial-none.jls with interleave mode "none" was not correctly converted from .ppm to .jls. This was fixed (charls-image-test tool)
2) The javascript conversion from 16 bit pixels to a web image was not correct (created a PR for it)
3) By default RGB images would be displayed as signed, which would cause strange colors (included in the PR).

I added a link to your repository and demo in the README.md of CharLS.
